coresimd/x86/cpuid.rs

```diff
@@ -86,26 +86,37 @@ pub fn has_cpuid() -> bool {
     }
     #[cfg(target_arch = "x86")]
     {
-        use coresimd::x86::{__readeflags, __writeeflags};
-
-        // On `x86` the `cpuid` instruction is not always available.
-        // This follows the approach indicated in:
-        // http://wiki.osdev.org/CPUID#Checking_CPUID_availability
         unsafe {
-            // Read EFLAGS:
-            let eflags: u32 = __readeflags();
-
-            // Invert the ID bit in EFLAGS:
-            let eflags_mod: u32 = eflags ^ 0x0020_0000;
-
-            // Store the modified EFLAGS (ID bit may or may not be inverted)
-            __writeeflags(eflags_mod);
-
-            // Read EFLAGS again:
-            let eflags_after: u32 = __readeflags();
-
-            // Check if the ID bit changed:
-            eflags_after != eflags
+            // On `x86` the `cpuid` instruction is not always available.
+            // This follows the approach indicated in:
+            // http://wiki.osdev.org/CPUID#Checking_CPUID_availability
+            // https://software.intel.com/en-us/articles/using-cpuid-to-detect-the-presence-of-sse-41-and-sse-42-instruction-sets/
+            // which detects whether `cpuid` is available by checking whether the 21th bit of the EFLAGS register is modifiable or not.
+            // If it is, then `cpuid` is available.
+            let eax: i32;
+            asm!(r#"
+                push %ecx
+                pushfd
+                pushfd
+                pop %eax
+                mov %ecx, %eax
+                xor %eax, 0x200000
+                push %eax
+                popfd
+                pushfd
+                pop %eax
+                xor %eax, %ecx
+                shrd %eax, 21
+                popfd
```

Comment:
Either this line (along with the corresponding `pushfd` is going to break `test_has_cpuid` which is testing that eflags has changed or, more likely, that test was already totally broken and should just be deleted. In fact, since the comparison is going to potentially modify `eflags`, I don't think that test makes any sense at all.

Reply:
`test_has_cpuid` just checks that the function returns `true`, there is a new test that checks that `has_cpuid` is indempotent. Maybe both tests should also check that the EFLAGS register wasn't modified, what do you think?

Reply:
sorry, I think I didn't fully gotten what you meant on the first read. I just changed the test to verify that calling `has_cpuid` does not modify the `EFLAGS` since I think it should be idempotent.

Reply:
``` rust
    fn test_has_cpuid() {
        unsafe {
            let before = __readeflags();

            if cpuid::has_cpuid() {
                assert!(before != __readeflags());
            } else {
                assert!(before == __readeflags());
            }
        }
    }
```
is testing that `eflags` changes if `cpuid` is supported and doesn't change if it's not. But the second and third `__readeflags()` happen after performing a comparison and a conditional jump. The comparison (probably a `test` instruction in this case) is going to set/reset bits in `eflags` which makes the comparison with `before` meaningless.